spec/ssh_scan/rule_engine_spec.rb

```diff
@@ -0,0 +1,125 @@
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_eq", 2.0, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_eq", 1.99, result)).to be false
+    end
+
+    it "neq should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_neq", 1.99, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_neq", 2.0, result)).to be false
+    end
+
+    it "gt should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 1.3, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gt", 2.0, result)).to be false
+    end
+
+    it "gte should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gte", 1.3, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gte", 2.0, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_gte", 2.1, result)).to be false
+    end
+
+    it "lt should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lt", 2.1, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lt", 2.0, result)).to be false
+    end
+
+    it "lte should work for numerics" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 2.1, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 2.0, result)).to be true
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_version_lte", 1.3, result)).to be false
+    end
+
+    it "should ignore attributes not in results" do
+      result = {:ssh_version => 2.0}
+      expect(SSHScan::RuleEngine.matchSingleRule("ssh_lib_eq", "openssh", result)).to be true
+    end
+  end
+
+  # Note: inputs can be given in a hash as well
+  context "parse conditional boolean expressions" do
+    it "simple true/false" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x", "x" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x", "x" => false)).to be false
+    end
+
+    it "should work for NOT expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("!x", "x" => false)).to be true
+    end
+
+    it "should work for AND expressions" do
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => true)).to be true
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => true, "y" => false)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => true)).to be false
+      expect(SSHScan::RuleEngine.parseBooleanRule("x && y", "x" => false, "y" => false)).to be false
```

Comment:
I like that you're going through all the cases here, having complete coverage can give you a lot more confidence in the code not having logic/boundary bugs.